Init

gradden opened the pull request.

gradden committed:
Uploading base game

gradden committed:
Alapleírás hozzáadása

gradden committed:
Frissített leírás

Nagyobb betűtípusú alcímek

oliverosz commented on main.cpp:
```diff
@@ -0,0 +1,69 @@
+    //A játék lefolyása
+    while(!warrior2->isDead() || !warrior1->isDead()){
+        
+        warrior1->damaging(warrior2);
+        cout << warrior1->getName() << " HP: " << warrior1->getHP() << endl;
+        cout << warrior2->getName() << " HP: " << warrior2->getHP() << endl;
+        
+        warrior2->damaging(warrior1);
+        cout << warrior1->getName() << " HP: " << warrior1->getHP() << endl;
+        cout << warrior2->getName() << " HP: " << warrior2->getHP() << endl;
```
Ez nem jó, ha warrior2 meghalt, akkor már ne üssön vissza. Látom, hogy a damaging ellenőrzi a saját HP-t, de az extra kiírás miatt akkor se jó a kimenet.

oliverosz replied:
A kiírásra meg legyen egy függvény, vagy << operátor, ne ctrl+c, ctrl+v.

gradden replied:
@oliverosz szintúgy ha ezeket a Hero.cpp-ben külön függvényben megoldható a kiiratás, vagy main.cpp-ben inkább?

oliverosz replied:
Ha a Hero.cpp-ben van a függvény, az nem látszik a main.cpp-ből, de talán erre te is rájöttél közben.

oliverosz commented on main.cpp:
```diff
@@ -0,0 +1,69 @@
+        if(warrior1->isDead()){
+            cout << warrior1->getName() << " is dead. " << warrior2->getName() << " wins. " << endl;
+            break;
+        }
```
Ez logikusabb helyen lenne a ciklus után.

oliverosz commented on main.cpp:
```diff
@@ -0,0 +1,69 @@
+            warrior1 = new Hero("Superman", 150, 20);
+            warrior2 = new Hero("Batman", 200, 30);
+        }
+    } catch (exception e) {
```
Exceptiont referenciaként kapunk el, nem érték szerint. Válaszban várok 2 okot, hogy miért úgy kell elkapni.

gradden replied:
@oliverosz igen ezt valóban elnéztem. Az indoka az, hogy más értékeket fog dobni az 'e' változó. Megváltozik a hibaüzenet, ha felülírom a változót és nem referálok rá.

oliverosz replied:
1. A változó nem tud dobni semmit sem, így nem tudom értelmezni, hogy mi akart lenni az első indok. Erős fantáziával talán gondolhatok arra, hogy a "más értékek" alatt más típusú exceptiont értesz, ami dobódik, és ha referencia, akkor upcastolódik exception típusra.
2. Ha felülírod, akkor mindenképp megváltozik a tartalma, akár referencia, akár nem.
A másik ok ugyanaz, amiért függvényparaméterben is preferáljuk a ref. szerinti átadást, hogy ne másolódjon feleslegesen az objektum.

oliverosz commented on Hero.h:
```diff
@@ -0,0 +1,28 @@
+private:
+    std::string name;
+    int hp;
+    int dmg;
```
A hp változik, de ami nem, az legyen const!

oliverosz commented on Hero.h:
```diff
@@ -0,0 +1,28 @@
+    
+public:
+    //Ctor, Dtor
+    Hero(const std::string, const int, const int);
```
Az egyszerű paramétereknél felesleges a const, a stringnél viszont még hiányolok egy referenciát. Igazából az a lényeg, nem a const, hogy ne másolódjon átadáskor.

oliverosz commented on Hero.h:
```diff
@@ -0,0 +1,28 @@
+    Hero(const std::string, const int, const int);
+    ~Hero();
+    //Karakter Alapadatok, Get-Set - Név, HP, DMG
+    std::string getName();
```
A string adattagot visszaadáskor sem másoljuk, hanem const refként adjuk vissza.

oliverosz commented on Hero.h:
```diff
@@ -0,0 +1,28 @@
+    ~Hero();
+    //Karakter Alapadatok, Get-Set - Név, HP, DMG
+    std::string getName();
+    void setName(std::string name);
```
const ref paraméter

oliverosz commented on Hero.h:
```diff
@@ -0,0 +1,28 @@
+    int getHP();
+    void setHP(int hp);
+    int getDmg();
+    void setDmg(int dmg);
```
Ilyen funkció nincs, ami változtatja a sebzést. Amíg valamire nincs szükség, addig ne legyen feleslegesen rá interfész. Ugyanígy, a setHP és setName sem megengedett művelet. A getterek pedig legyenek constok.

oliverosz replied:
Ismétlem: a getterek legyenek constok. Meg úgy általában minden metódus, ami const lehet.

oliverosz commented on Hero.cpp:
```diff
@@ -0,0 +1,60 @@
+#include <string>
```
Itt nem kell újra include-olni a stringet, az már be van a Hero.h által. És általában az osztály headerjét szokás elsőként include-olni.

oliverosz commented on Hero.cpp:
```diff
@@ -0,0 +1,60 @@
+Hero::Hero(const std::string name, const int hp, const int dmg) : name(name), hp(hp), dmg(dmg){
+    this->name = name;
+    this->hp = hp;
+    this->dmg = dmg;
```
Ez itt miért kell?

oliverosz commented on Hero.cpp:
```diff
@@ -0,0 +1,60 @@
+    
+}
+
+Hero::~Hero(){}
```
Ha üres, akkor minek van definiálva?

oliverosz commented on Hero.cpp:
```diff
@@ -0,0 +1,60 @@
+
+void Hero::damaging(Hero *enemy){
+    if(hp > 0){
+        std::cout << name << " is damaging "<< enemy->name << " with: " << dmg << " dmg" << std::endl;
```
Jobb lenne a kiíratást kivenni a metódusból. Egy metódusnak feladata legyen! (single responsibility)

gradden replied:
@oliverosz az megoldható, hogy a Hero.cpp-be külön függvényt hozok létre a kiiratásoknak? Vagy azokat célszerűbb a main.cpp-ben megcsinálni külön függvénybe?

gradden committed:
Kijavított hibák

Javított hibák:
- extra kiírás, ha valamelyik karakter meghalt
- 2 metódus hozzáadva, amely az outputokat kezeli
- "exception e" javítása
- a 'Hero.h' headerben javítva az adattagok. (const, reference, ctor stb..)
- a felesleges függvények törölve a 'Head.cpp' fájlban, illetve a metódusok javítása

gradden commented:
@hegyhati @oliverosz javítottuk a hibákat!

oliverosz commented on Hero.h:
```diff
@@ -0,0 +1,24 @@
+#ifndef Hero_h
+#define Hero_h
+#include <string>
+#include <iostream>
```
Az iostream hol van használva ebben a fájlban?

gradden replied:
@oliverosz a main.cpp-ben használódik, gondoltam egybe szedem ezeket, és a main.cpp-ben csak a header fájlt kell includeolni.

oliverosz replied:
Hát ez nem jó gondolat. Minden fájlban az legyen include-olva, amire abban a fájlban szükség van.

oliverosz commented on Hero.cpp:
```diff
@@ -0,0 +1,41 @@
+    
+}
+
+const int& Hero::getHP(){
+    return hp;
+    
+}
+
+const int& Hero::getDmg(){
```
Csak az összetett adattagoknál kell referenciát visszaadni. Az egyszerű típusoknál nem gond a másolás, ott a referencia a rosszabb megoldás.

oliverosz commented on main.cpp:
```diff
@@ -0,0 +1,72 @@
+    //A játék lefolyása
+    while(!warrior2->isDead() && !warrior1->isDead()){
+        
+        if(!warrior1->isDead()){
+            warrior1->damaging(warrior2);
+            std::cout << warrior1->getName() << " is damaging "<< warrior2->getName() << " with: " << warrior1->getDmg() << " dmg" << std::endl;
+            getBattleStatus(*warrior1, *warrior2);
+        }
+        
+        
+        if(!warrior2->isDead()){
+            warrior2->damaging(warrior1);
+            std::cout << warrior1->getName() << " is damaging "<< warrior2->getName() << " with: " << warrior1->getDmg() << " dmg" << std::endl;
+            getBattleStatus(*warrior1, *warrior2);
+        }
```
Most már a kimenet helyes, de nem igaz, hogy nem zavar a redundancia ebben a megoldásban... [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

oliverosz commented on main.cpp:
```diff
@@ -0,0 +1,72 @@
+    << "Damage: " << warrior.getDmg() << std::endl << std::endl;
+}
+
+const void getBattleStatus(Hero& warrior1, Hero& warrior2){
+    std::cout << warrior1.getName() << " HP: " << warrior1.getHP() << std::endl;
+    std::cout << warrior2.getName() << " HP: " << warrior2.getHP() << std::endl;
```
1. Mit jelent a const void visszatérési érték?
2. Ezt a 2 sort kéne kiváltani gy függvénnyel, mert majdnem ugyanazok. És nem ezt a 2 sort beletenni egy függvénybe, mert ezt úgyis csak egyszer kellene végrehajtani, ha a main loopból ki lenne szedve a redundancia.
3. A paraméterek miért nem constok?

gradden committed:
Hibák javítása 2

- redundancia fokozása
- getterek constra rakása

gradden commented:
@oliverosz Javítottuk

oliverosz commented on main.cpp:
```diff
@@ -40,26 +34,18 @@ int main(int argc, const char * argv[]) {
-            getBattleStatus(*warrior1, *warrior2);
-        }
-        
+        warrior1->damaging(warrior2);
+        std::cout << warrior1->getName() << "->"<< warrior2->getName() << std::endl;
+        warrior2->damaging(warrior1);
+        std::cout << warrior2->getName() << "->"<< warrior1->getName() << std::endl;
```
Most gyakorlatilag visszatértünk az első verzióra... Mindegy, nem egy szép megoldás, de elfogadható. A többi rész már rendben van.

oliverosz reviewed (approved):
@hegyhati merge
